### ISSUE TITLE
Singularity Clearing supplementary GIDs preventing binds

### ISSUE THREAD

**Issue by sabbene** [apptainer/singularity#1068]
## Version of Singularity:
```
$ /tmp/singularity-2.4/bin/singularity --version
2.4-dist
```

### Expected behavior

Singularity to bind paths that a user has permissions to access.

### Actual behavior

Singularity is not binding to paths where a dir in the hierarchy is owned by a group that is not their primary group.

### Steps to reproduce behavior

Write here.

Lets use this tree and users for an example:

```
$ ls -ld /home/scratch/
drwxrwxrwx 39 user1 group1 192512 Oct 13 16:41 /home/scratch


$ ls -ld /home/scratch/test
drwxrwx--- 3 user1 group2 4096 Oct 13 16:41 /home/scratch/test


$ ls -ld /home/scratch/test/test2
drwxr-xr-x 2 user1 group1 4096 Oct 13 16:41 /home/scratch/test/test2


$ id
uid=789(user1) gid=123(group1) groups=123(group1),456(group2)
```

When I use singularity to bind to /home/scratch or /home/scratch/test everything works fine, but if I try to bind to /home/scratch/test/test2 i get a permission denied error, even though the user1 user has access to the dir.  From looking at the --debug output when I recieve this error it looks like singularity drops all the user's groups but the primary one.  

```
$ /tmp/singularity-2.4/bin/singularity --debug shell --bind /home/scratch/test/test2:/home/test docker://centos:latest
...
...
DEBUG   [U=9035,P=33839]   singularity_registry_get()                Returning value from registry: 'OVERLAYFS_ENABLED' = '1'
VERBOSE [U=9035,P=33839]   singularity_runtime_mount_userbinds()     Creating bind directory on overlay file system: /home/test
DEBUG   [U=9035,P=33839]   s_mkpath()                                Creating directory: /tmp/singularity-2.4/var/singularity/mnt/final/home/test
DEBUG   [U=9035,P=33839]   singularity_priv_escalate()               Temporarily escalating privileges (U=9035)
DEBUG   [U=0,P=33839]      singularity_priv_escalate()               Clearing supplementary GIDs.
VERBOSE [U=0,P=33839]      singularity_runtime_mount_userbinds()     Binding '/home/scratch/test/test2' to '/tmp/singularity-2.4/var/singularity/mnt/final//home/test'
ERROR   [U=0,P=33839]      singularity_runtime_mount_userbinds()     There was an error binding the path /home/scratch/test/test2: Permission denied
ABORT   [U=0,P=33839]      singularity_runtime_mount_userbinds()     Retval = 255
...
```

Please let me know if any additional information or testing is necessary,
 Stephen Abbene

**Comment by cclerget** [comment]
@sabbene I can't reproduce the bug by setting same permissions to directories. It's weird you got permission denied while the operation is ran with root privilege. Can you post full debug log please ?

**Comment by sabbene** [comment]
Are you testing on NFS that does not have the no squash root setting enabled?

I can provide the full debug log later today. 

**EDIT** 

From our conversation on slack you no longer need the debug output.

A list of things that will not work:

newgrp
singularity -u
chmod o+x  (due to infosec concerns)

**Comment by cclerget** [comment]
@sabbene Fix submitted, you can test from branch https://github.com/cclerget/singularity/tree/development-nfs-root-squash-fix.

EDIT:
Fix will probably changes, but if you can validate it works for you in a first time

**Comment by sabbene** [comment]
Did I grab the wrong version?  I am still seeing issues binding to the path.

```
$ ./singularity --version
2.4-master.g81bf030
```

```
$ ./singularity --debug shell --bind /home/scratch/test/test2:/home/test docker://centos:latest
DEBUG   [U=9035,P=64642]   singularity_runtime_mount_userbinds()     Checking for environment variable 'SINGULARITY_BINDPATH'
DEBUG   [U=9035,P=64642]   singularity_registry_get()                Returning value from registry: 'BINDPATH' = ',/home/scratch/test/test2:/home/test'
DEBUG   [U=9035,P=64642]   singularity_runtime_mount_userbinds()     Checking for 'user bind control' in config
DEBUG   [U=9035,P=64642]   singularity_config_get_bool_char_impl()   Called singularity_config_get_bool(user bind control, yes)
DEBUG   [U=9035,P=64642]   singularity_config_get_value_impl()       Returning configuration value user bind control='yes'
DEBUG   [U=9035,P=64642]   singularity_config_get_bool_char_impl()   Return singularity_config_get_bool(user bind control, yes) = 1
DEBUG   [U=9035,P=64642]   singularity_runtime_mount_userbinds()     Parsing SINGULARITY_BINDPATH for user-specified bind mounts.
DEBUG   [U=9035,P=64642]   singularity_runtime_mount_userbinds()     Found bind: /home/scratch/test/test2 -> container:/home/test
DEBUG   [U=9035,P=64642]   singularity_runtime_mount_userbinds()     Checking if bind point is already mounted: /home/test
DEBUG   [U=9035,P=64642]   check_mounted()                           Opening /proc/mounts
DEBUG   [U=9035,P=64642]   singularity_registry_get()                Returning value from registry: 'OVERLAYFS_ENABLED' = '1'
VERBOSE [U=9035,P=64642]   singularity_runtime_mount_userbinds()     Creating bind directory on overlay file system: /home/test
DEBUG   [U=9035,P=64642]   s_mkpath()                                Creating directory: /tmp/singularity-2.4-hotfix/var/singularity/mnt/final/home/test
DEBUG   [U=9035,P=64642]   singularity_priv_escalate()               Temporarily escalating privileges (U=9035)
DEBUG   [U=0,P=64642]      singularity_priv_escalate()               Clearing supplementary GIDs.
VERBOSE [U=0,P=64642]      singularity_runtime_mount_userbinds()     Binding '/home/scratch/test/test2' to '/tmp/singularity-2.4-hotfix/var/singularity/mnt/final//home/test'
ERROR   [U=0,P=64642]      singularity_runtime_mount_userbinds()     There was an error binding the path /home/scratch/test/test2: Permission denied
ABORT   [U=0,P=64642]      singularity_runtime_mount_userbinds()     Retval = 255
```

**Comment by sabbene** [comment]
I checked out the wrong commit.  It looks like it is working now!! Thank you so much.

```
$ ./singularity --version
2.4-HEAD.geb6829c
```
```
$ ./singularity exec --bind /home/scratch/test/test2:/home/test docker://centos:latest 'ls /home/test'
Docker image path: index.docker.io/library/centos:latest
Cache folder set to /home/test/.singularity/docker
Creating container runtime...
foo

$ echo $?
0
```

**Comment by GodloveD** [comment]
Yay!  Thanks @cclerget!  Are we good to close this issue?

**Comment by cclerget** [comment]
Yes @GodloveD. But need to discuss about implementation of PR #1100. Use of ld linker --wrap option or a singularity_mount function wrapper ?

**Comment by dtrudg** [comment]
Closing as issue is fixed, and referenced #1100 is also merged.
